python/tests/unittests/test_SpinOperator.py

```diff
@@ -152,6 +155,15 @@ def test_spin_op_vqe():
     assert want_string == hamiltonian.to_string()
     assert want_string == str(hamiltonian)
 
+def test_matrix():
+    """
+    Test that the `cudaq.SpinOperator` can produce its matrix representation 
+    and that we can use that matrix with standard python packages like numpy.
+    """
+    hamiltonian = 5.907 - 2.1433 * spin.x(0) * spin.x(1) - 2.1433 * spin.y(
+        0) * spin.y(1) + .21829 * spin.z(0) - 6.125 * spin.z(1)
+    mat = hamiltonian.to_matrix()
+    assert_close(-1.74, np.linalg.eigvals(mat)[0], 1e-2)
```

Comment:
This is something for a later PR but I've had an issue with the state vector PR on something similar to this. Our `ComplexMatrix` (and `State`) classes are functional as matrices/vectors in python, but we can't do any operations between them and numpy ndarrays because we don't have a type conversion defined. 

Something I'm playing around with for the state class is defining a `to_numpy` method that creates the state vec/dm as an eigen matrix, then uses pybinds built-in type conversion to return that eigen matrix as a numpy ndarray. Like this 
```
.def("to_numpy", [](){ 
   // build up `matrix` as an Eigen::MatrixXcd
   return matrix;
}, py::return_value_policy::reference_internal) 
```

The `reference_internal` then ties the life of the returned np.ndarray to the life of the class (`self`). 

We may consider doing something similar for the `ComplexMatrix` class such that someone could easily convert it to a numpy.ndarray and then be able to use it with numpy functions. Here, that would let us define the `want_matrix` as a numpy array and assert that 
```
assert np.allclose(want_matrix, mat.to_numpy())
```